src/core/main/common/FreezeResolver.ts

```diff
@@ -0,0 +1,533 @@
+      if (status.type === "initialized") {
+        for (const segment of status.value.getLastKnownInventory()) {
+          const { representation } = segment.infos;
+          if (representation.decipherable === false) {
+            log.warn("FR: we have undecipherable segments left in the buffer, reloading");
+            this._decipherabilityFreezeStartingTimestamp = null;
+            this._ignoreFreezeUntil = now + 6000;
+            return { type: "reload", value: null };
+          } else if (representation.contentProtections !== undefined) {
+            isClear = false;
+            if (representation.decipherable !== true) {
+              hasOnlyDecipherableSegments = false;
+            }
+          }
+        }
+      }
+    }
+
+    if (shouldHandleDecipherabilityFreeze && !isClear && hasOnlyDecipherableSegments) {
+      log.warn(
+        "FR: we are frozen despite only having decipherable " +
+          "segments left in the buffer, reloading",
+      );
+      this._decipherabilityFreezeStartingTimestamp = null;
+      this._ignoreFreezeUntil = now + 6000;
+      return { type: "reload", value: null };
+    }
+    return null;
+  }
+
+  /**
+   * This method should only be called if a "flush" strategy has recently be
+   * taken to try to unfreeze playback yet playback is still frozen.
+   *
+   * It considers the current played content and returns a more-involved
+   * unfreezing strategy (most often reload-related) to try to unfree playback.
+   * @param {number} freezingPosition - The playback position at which we're
+   * currently frozen.
+   * @returns {Object}
+   */
+  private _getStrategyIfFlushingFails(freezingPosition: number): IFreezeResolution {
+    log.warn(
+      "FR: A recent flush seemed to have no effect on freeze, checking for transitions",
+    );
+
+    /** Contains Representation we might want to avoid after the following algorithm */
+    const toAvoid = [];
+
+    for (const ttype of ["audio", "video"] as const) {
+      const segmentList = this._lastSegmentInfo[ttype];
+      if (segmentList.length === 0) {
+        // There's no buffered segment for that type, go to next type
+        continue;
+      }
+
+      /** Played history information on the current segment we're stuck on. */
+      let currentSegmentEntry = segmentList[segmentList.length - 1];
+      if (currentSegmentEntry.segment === null) {
+        // No segment currently played for that given type, go to next type
+        continue;
+      }
+
+      /** Metadata on the segment currently being played. */
+      const currentSegment = currentSegmentEntry.segment;
+
+      /**
+       * Set to the first previous segment which is linked to a different
+       * Representation.
+       */
+      let previousRepresentationEntry: IPlayedHistoryEntry | undefined;
+
+      // Now find `previousRepresentationEntry` and `currentSegmentEntry`.
+      for (let i = segmentList.length - 2; i >= 0; i--) {
+        const segment = segmentList[i];
```

Comment:
what if segmentList.length := 1 ?
I'm not sure we are checking this correctly

Reply:
We want to find the previous segment I think here. It makes no sense if there's only one entry in history and in that case `currentSegmentEntry` is already the earliest reference to the current segment, so I think that loop is OK

Reply:
I mean `segmentList[i]` will be `undefined` if `i` is below 0.
Next line `segment.segment` will throw an error if segment is `undefined`.

Reply:
Well condition `i >= 0;` guarantees that `i` is positive so there will be no error here